src/packages/client/src/runtime/getPrismaClient.ts

```diff
@@ -915,45 +915,42 @@ new PrismaClient({
      * @param middlewareIndex
      * @returns
      */
-    private _request(
+    private async _request(
       internalParams: InternalRequestParams,
-      middlewareIndex = 0,
     ): Promise<any> {
       try {
-        // in this recursion, we check for our terminating condition
-        const middleware = this._middlewares.query.get(middlewareIndex)
+        let index = -1
         // async scope https://github.com/prisma/prisma/issues/3148
         const resource = new AsyncResource('prisma-client-request')
+        // make sure that we don't leak extra properties to users
+        const params: QueryMiddlewareParams = {
+          args: internalParams.args,
+          dataPath: internalParams.dataPath,
+          runInTransaction: internalParams.runInTransaction,
+          action: internalParams.action,
+          model: internalParams.model,
+        }
+
+        // prepare recursive fn that will pipe params through middlewares
+        const consumer = (changedParams: QueryMiddlewareParams) => {
+          // if this `next` was called and there's some more middlewares
+          const nextMiddleware = this._middlewares.query.get(++index)
 
-        if (middleware) {
-          // make sure that we don't leak extra properties to users
-          const params: QueryMiddlewareParams = {
-            args: internalParams.args,
-            dataPath: internalParams.dataPath,
-            runInTransaction: internalParams.runInTransaction,
-            action: internalParams.action,
-            model: internalParams.model,
+          if (nextMiddleware) {
+            // we pass the modfied params down to the next one, & repeat
+            return nextMiddleware(changedParams, consumer)
           }
 
-          return resource.runInAsyncScope(() => {
-            // call the middleware of the user & get their changes
-            return middleware(params, (changedParams) => {
-              // this middleware returns the value of the next one 🐛
-              return this._request(
-                {
-                  ...internalParams,
-                  ...changedParams,
-                },
-                ++middlewareIndex,
-              ) // recursion happens over here
-            })
-          })
+          const changedInternalParams = { ...internalParams, ...params }
+
+          // TODO remove this, because transactionId should be passed?
+          if (index > 0) delete changedInternalParams['transactionId']
```

Comment:
TODO for now?

Reply:
Yes

Reply:
We will need to remove this, because middlewares with transactions makes it not to be executed as a regular transaction (removes `transactionId`)

Reply:
Can we make an issue out of this so we can remember and prioritize? Thanks.

Reply:
This is preventing the transactions from rolling back. If enabled, transactions do rollback, but introduce those bugs we've seen on studio. https://github.com/prisma/prisma/issues/6705#issuecomment-861055614